### PR TITLE
Memory leak and invalid use fixes

### LINKED ISSUE #2146: Can't Train a Normal CNN.
#### Issue description

When I try to train the following cnn, it only runs when I include the line 

`arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");`

However, if I comment out this line, the code no longer runs (both cases compile. but in this second case doesn't run - "Segmentation Fault"). It doesn't make sense to me why this happens.

#### Your environment

 * version of mlpack: 3.2.2
 * operating system: ubuntu 18.04
 * compiler: g++ 7.4.0
 * version of dependencies (Boost/Armadillo): libboost-1.65.1, armadillo-9.800.3

#### Steps to reproduce

 Here's the code to reproduce the issue: 

    #include <mlpack/core.hpp>
    #include <mlpack/core/data/split_data.hpp>
    #include <mlpack/methods/ann/layer/layer.hpp>
    #include <mlpack/methods/ann/ffn.hpp>
    #include <ensmallen.hpp>

    using namespace mlpack;
    using namespace mlpack::ann;
    using namespace arma;
    using namespace std;
    using namespace ens;

    int main()
    {
        double RATIO = 0.1;
        int CYCLES = 10;
        const int ITERATIONS_PER_CYCLE = 1;
        double STEP_SIZE = 10e-3;
        int BATCH_SIZE = 1;

        // Reading in image.
        cout << "Reading data ..." << endl;
        arma::Mat<unsigned char> matrix;
        data::ImageInfo info;
        data::Load("saitama.png", matrix, info, false, true);

        FFN<NegativeLogLikelihood<>, RandomInitialization> model;

        model.Add<Convolution<> >(info.Channels(), 6, 5, 5, 1, 1, 0, 0, info.Width(), info.Height());
        model.Add<ReLULayer<> >();
        model.Add<MaxPooling<> >(2, 2, 2, 2, true);
        model.Add<Linear<> >(62*62*6, 2);
        model.Add<LogSoftMax<> >();
        cout << "Training ..." << endl;

        // Setting parameters Stochastic Gradient Descent (SGD) optimizer.
        SGD<AdamUpdate> optimizer(
          STEP_SIZE,
          BATCH_SIZE,
          ITERATIONS_PER_CYCLE,
          1e-8,
          true,
          AdamUpdate(1e-8, 0.9, 0.999));

        arma::mat trainY;
        trainY << 1;// << 0;
        arma::conv_to<arma::mat>::from(matrix).print("Print Image: \n");
        trainY.print("Train label: \n");
        cout << trainY.n_rows << " " << trainY.n_cols;

        for (int i = 1; i <= CYCLES; i++)
        {
          model.Train(arma::conv_to<arma::mat>::from(matrix), trainY, optimizer, PrintLoss(), ProgressBar());
        }
        return 0;
    }

(The image size I used is 128 x 128).

compiled using: 
```g++ cnn.cpp -o cnn `pkg-config --cflags --libs mlpack armadillo` -fopenmp -DHAS_STB```
#### Expected behavior

When the line is commented out, the network should still train. The code should run irrespective of whether the line in question is commented out or not.

#### Actual behavior

The code runs only when the line is NOT commented out.

### PR DESCRIPTION
**Pull request by rcurtin**
Inspired by #2314, #2310, #2300, and others, I built mlpack with `-fsanitize=address` and ran `mlpack_test`.  The results were... noisy.  But I took some time and I fixed everything (except for one, detailed later).

 * All of the ANN layers are properly freed now; it took a while to get this write, and `DeleteVisitor` will now delete internally-held layers when the layer has a `Model()` method.  This may solve some issues people were having.

 * CNN layers don't make an alias of the input from `Forward()` anymore.  This should fix #2146 .

 * Test code properly frees things that it allocates.  There were some tricky places where deallocation had been overlooked.

 * Bindings handle exceptions properly and deallocate any allocated memory when an exception is thrown.  It makes the code a bit uglier, but it's safer.

There is one bug I didn't solve, which surfaces in the reinforcement learning code, in `q_learning_impl.hpp`:

```
     // Update target network
     if (totalSteps % config.TargetNetworkSyncInterval() == 0)
      targetNetwork = learningNetwork;
```

This code actually copies an `FFN`, and each individual layer will require a copy constructor to fix that memory leak.  That's the same issue as in #2314, and anyway, that isn't solved here.  But everything else in the tests is.

I do think it's important that we merge this before a release, otherwise we'll probably get lots of segfault and invalid memory access reports. :)

### CODE REVIEW

File: src/mlpack/methods/ann/layer/add_merge_impl.hpp [hunk 1]
```diff
@@ -150,7 +159,15 @@ void AddMerge<InputDataType, OutputDataType, CustomLayers...>::serialize(
   ar & BOOST_SERIALIZATION_NVP(network);
   ar & BOOST_SERIALIZATION_NVP(model);
   ar & BOOST_SERIALIZATION_NVP(run);
-  ar & BOOST_SERIALIZATION_NVP(ownsLayer);
+
+  if (version >= 1)
```
Comment (birm) [inline src/mlpack/methods/ann/layer/add_merge_impl.hpp]: guess you have to uncomment version in the definition :)

File: src/mlpack/methods/ann/layer/add_merge_impl.hpp [hunk 2]
```diff
@@ -150,7 +159,15 @@ void AddMerge<InputDataType, OutputDataType, CustomLayers...>::serialize(
...
+  {
```
Comment (favre49) [inline src/mlpack/methods/ann/layer/add_merge_impl.hpp]: We can remove the braces here

File: src/mlpack/methods/ann/visitor/delete_visitor_impl.hpp [hunk 3]
```diff
@@ -20,12 +20,30 @@ namespace ann {
...
 
+template<typename LayerType>
+inline typename std::enable_if<
+    HasModelCheck<LayerType>::value, void>::type
+DeleteVisitor::operator()(LayerType* layer) const
+{
+  if (layer)
+  {
+    for (size_t i = 0; i < layer->Model().size(); ++i)
+    {
```
Comment (favre49) [inline src/mlpack/methods/ann/visitor/delete_visitor_impl.hpp]: We can remove the braces here too

File: src/mlpack/methods/hoeffding_trees/hoeffding_tree_model.hpp [hunk 4]
```diff
@@ -188,30 +188,18 @@ class HoeffdingTreeModel
     data::DatasetInfo info;
     if (type == GINI_HOEFFDING)
     {
```
Comment (favre49) [inline src/mlpack/methods/hoeffding_trees/hoeffding_tree_model.hpp]: Here too

File: src/mlpack/methods/ann/brnn_impl.hpp [hunk 5]
```diff
@@ -60,6 +60,22 @@ BRNN<OutputLayerType, MergeLayerType, MergeOutputType,
...
+template<typename OutputLayerType, typename MergeLayerType,
+         typename MergeOutputType, typename InitializationRuleType,
+         typename... CustomLayers>
+BRNN<OutputLayerType, MergeLayerType, MergeOutputType,
+    InitializationRuleType, CustomLayers...>::~BRNN()
+{
+  // Remove mergeLayer from the forward and backward RNNs so it doesn't get
+  // deleted.  This assumes that mergeLayer is the last layer!
+  forwardRNN.network.pop_back();
+  backwardRNN.network.pop_back();
```
Comment (saksham189) [inline src/mlpack/methods/ann/brnn_impl.hpp]: This doesn't look right. We never add the mergeLayer to the forward and backward RNN but the last layer of the RNN's to the merge layer if you look [here](https://github.com/mlpack/mlpack/blob/3199df34fef4eb10ba26487561d7d678b04e6fe9/src/mlpack/methods/ann/brnn_impl.hpp#L635-L642).
I guess we only need change the comment though. If reset is true then we should pop the last layers from the RNN's to avoid deleting twice otherwise we should simply delete the mergeLayer

Comment (rcurtin) [reply src/mlpack/methods/ann/brnn_impl.hpp]: Oh, you're right!  I read it backwards.  In any case, yeah, removing the last layer from the forward and backward RNNs is sufficient to prevent a memory leak.  I've updated the comment in c42b9f767. :+1:

**Comment by rcurtin** [comment]
Thanks @favre49, most of the time was actually just waiting on the build and seeing the results. :)  I removed all the unnecessary braces in b1e51468f, now let's see if it still builds clean. :+1:

@birm yeah, there is some more info on the HDF5 test in #2320.

**Comment by zoq** [comment]
Thanks for spending so many ours on the issues, I'll take a look at the changes later today.

File: src/mlpack/methods/ann/brnn_impl.hpp [hunk 6]
```diff
@@ -60,6 +60,23 @@ BRNN<OutputLayerType, MergeLayerType, MergeOutputType,
...
 
+template<typename OutputLayerType, typename MergeLayerType,
+         typename MergeOutputType, typename InitializationRuleType,
+         typename... CustomLayers>
+BRNN<OutputLayerType, MergeLayerType, MergeOutputType,
+    InitializationRuleType, CustomLayers...>::~BRNN()
+{
+  // Remove the last layers from the forward and backward RNNs, as they are held
+  // in mergeLayer.  So, when we use DeleteVisitor with mergeLayer, those two
+  // layers will be properly (and not doubly) freed.
```
Comment (zoq) [inline src/mlpack/methods/ann/brnn_impl.hpp]: Ohh, nice!

**Comment by favre49** [comment]
I'll go ahead and merge this then.
